--- plugins/stardict/lib.cpp    (revision 454)
+++ plugins/stardict/lib.cpp    (revision 455)
@@ -1727,11 +1727,13 @@
                     oFuzzystruct[iMaxDistanceAt].iMatchWordDistance = iDistance;
                     // calc new iMaxDistance
                     iMaxDistance = iDistance;
+                    int tmpVal = iMaxDistance; //stupid workaround for gcc bug 44545
                     for (int j = 0; j < reslist_size; j++)
                     {
                         if (oFuzzystruct[j].iMatchWordDistance > iMaxDistance)
-                            iMaxDistance = oFuzzystruct[j].iMatchWordDistance;
+                            tmpVal = oFuzzystruct[j].iMatchWordDistance;
                     } // calc new iMaxDistance
+                    iMaxDistance = tmpVal; // end of stupid workaround
                 }   // add to list
             }   // find one
         }   // each word
