diff -urN psi_old/src/pluginhost.cpp psi_new/src/pluginhost.cpp
--- psi_old/src/pluginhost.cpp	2009-06-27 13:55:18.000000000 +0900
+++ psi_new/src/pluginhost.cpp	2009-06-30 11:20:52.000000000 +0900
@@ -365,13 +365,12 @@
  * this will call its processEvent() handler.
  * Handler may then modify the event and may cause the event to be
  * silently discarded.
- * TODO: modification doesn't work
  * 
  * \param account Identifier of the PsiAccount responsible
  * \param e Event XML
  * \return Continue processing the event; true if the stanza should be silently discarded.
  */
-bool PluginHost::processEvent(int account, const QDomElement& e)
+bool PluginHost::processEvent(int account, QDomElement& e)
 {
 	bool handled = false;
 	EventFilter *ef = qobject_cast<EventFilter*>(plugin_);
diff -urN psi_old/src/pluginhost.h psi_new/src/pluginhost.h
--- psi_old/src/pluginhost.h	2009-06-27 13:55:18.000000000 +0900
+++ psi_new/src/pluginhost.h	2009-06-30 11:21:32.000000000 +0900
@@ -53,7 +53,7 @@
 	bool incomingXml(int account, const QDomElement& e);
 
 	// for EventFilter
-	bool processEvent(int account, const QDomElement& e);
+	bool processEvent(int account, QDomElement& e);
 	bool processMessage(int account, const QString& jidFrom, const QString& body, const QString& subject);
 
 	// StanzaSendingHost
diff -urN psi_old/src/plugins/include/eventfilter.h psi_new/src/plugins/include/eventfilter.h
--- psi_old/src/plugins/include/eventfilter.h	2009-04-13 10:02:17.000000000 +0900
+++ psi_new/src/plugins/include/eventfilter.h	2009-06-30 11:22:16.000000000 +0900
@@ -10,7 +10,7 @@
 
 	// true = handled, don't pass to next handler
 
-    virtual bool processEvent(int account, const QDomElement& e) = 0;
+    virtual bool processEvent(int account, QDomElement& e) = 0;
 	virtual bool processMessage(int account, const QString& fromJid, const QString& body, const QString& subject) = 0;
 };
 
diff -urN psi_old/src/psiaccount.cpp psi_new/src/psiaccount.cpp
--- psi_old/src/psiaccount.cpp	2009-06-27 13:55:18.000000000 +0900
+++ psi_new/src/psiaccount.cpp	2009-06-30 11:19:31.000000000 +0900
@@ -3964,8 +3964,9 @@
 	if (PluginManager::instance()->processEvent(this, eXml)) {
 		delete e;
 		return;
+	} else {
+	  e->fromXml(psi(),this, &eXml);
 	}
-	//FIXME(KIS): must now cause the event to be recreated from this xml or such. Horrid. 	
 #endif
 	
 	if(d->acc.opt_log && activationType != FromXml) {
