diff -r 2f1f34ae1cc8 src/InputMapper.cpp
--- a/src/InputMapper.cpp	Thu Feb 16 01:36:22 2012 -0600
+++ b/src/InputMapper.cpp	Thu May 24 11:06:40 2012 +0200
@@ -323,9 +323,9 @@
 		"dance",
 		"Dance ", //Notice extra space at end
 		"LevelSix USB Pad (DDR638)",	// "DDR638" is the model number of the pad
-		AutoMappingEntry( 0, JOY_BUTTON_1,      DANCE_BUTTON_UP,	       false ),
+		AutoMappingEntry( 0, JOY_BUTTON_1,      DANCE_BUTTON_LEFT,	       false ),
 		AutoMappingEntry( 0, JOY_BUTTON_2,      DANCE_BUTTON_DOWN,      false ),
-		AutoMappingEntry( 0, JOY_BUTTON_3,      DANCE_BUTTON_LEFT,      false ),
+		AutoMappingEntry( 0, JOY_BUTTON_3,      DANCE_BUTTON_UP,      false ),
 		AutoMappingEntry( 0, JOY_BUTTON_4,      DANCE_BUTTON_RIGHT,     false ),
 		AutoMappingEntry( 0, JOY_BUTTON_7,      DANCE_BUTTON_UPRIGHT,   false ),
 		AutoMappingEntry( 0, JOY_BUTTON_8,      DANCE_BUTTON_UPLEFT,    false ),

